app/modules/softuart.c

```diff
@@ -0,0 +1,477 @@
+#include "lualib.h"
+#include "lauxlib.h"
+#include "platform.h"
+#include "module.h"
+#include "c_types.h"
+#include "c_string.h"
+
+#include "ets_sys.h"
+#include "osapi.h"
+#include "gpio.h"
+#include "os_type.h"
+
+#define SOFTUART_MAX_RX_BUFF 256
+
+//array of pointers to instances
+
+typedef struct softuart_pin_t {
+	uint8_t gpio_id;
+	uint32_t gpio_mux_name;
+	uint8_t gpio_func;
+} softuart_pin_t;
+
+typedef struct softuart_buffer_t {
+	char receive_buffer[SOFTUART_MAX_RX_BUFF]; 
+	uint8_t length;
+} softuart_buffer_t;
+
+typedef struct {
+    BOOL enabled;
+	softuart_pin_t pin_rx;
+	softuart_pin_t pin_tx;
+	softuart_buffer_t buffer; //TODO volatile ??? not needed as no read anymore
+    uint32_t baudrate;
+	uint16_t bit_time;
+    uint16_t need_len;
+    uint16_t end_char;
+    int uart_receive_rf;
+} Softuart;
+
+Softuart softuart = {false};
```

Comment:
Why using static variable instead of object-model?

Reply:
Thank you for your response. I would like to be as close as possible to the hardware uart module. What do you think of using ids instead of objects?

Reply:
I think IDs may be confusing here. Software UART may be used on any pair of pins, so...